### PR TITLE
#1364 - more flexible block fetching

### LINKED ISSUE #1364: Checkpoints break the syncing algorithm in `BlockStore`
Currently block store keeps a map of information about each peer:
https://github.com/Zilliqa/zq2/blob/e5b4d088efaff6e565569e90ca2aa8c5bc8e1d84/zilliqa/src/block_store.rs#L39

This information includes the greatest view of any block that we've seen from the peer:
https://github.com/Zilliqa/zq2/blob/e5b4d088efaff6e565569e90ca2aa8c5bc8e1d84/zilliqa/src/block_store.rs#L63

Currently this is only filled when we receive a proposal from a peer, but in future we want to add extra network messages by which peers can share their greatest view:
https://github.com/Zilliqa/zq2/blob/e5b4d088efaff6e565569e90ca2aa8c5bc8e1d84/zilliqa/src/block_store.rs#L262

We use this information to avoid making requests to peers for blocks they don't actually have:
https://github.com/Zilliqa/zq2/blob/e5b4d088efaff6e565569e90ca2aa8c5bc8e1d84/zilliqa/src/block_store.rs#L129

However, now we have checkpoints which mean nodes can be started from a checkpoint at block `n`, rather than always syncing from genesis. Validators can be started from checkpoints and propose blocks at views greater than `n`. This makes our assumption that a node who proposed `m` can provide us with blocks `0..=m` no longer true.

So we need to fix our block syncing algorithm to deal with this weaker property. I don't know how yet, lets discuss :)

### PR DESCRIPTION
**Pull request by rrw-zilliqa**
This PR is dependent on https://github.com/Zilliqa/zq2/pull/1494 and closes #1364 (though note that syncing from a checkpoint still doesn't work for me - this PR only fixes the fetching, not restoration per se)

There are docs in the PR; note that this generates a lot of tracing - which is what I found I needed to debug various failure cases (and is hence in the PR because I bet we will see more failures when it's merged).

I would ideally like (natch) to rip all this up and write a branch new block caching infrastructure which, for example, explicitly calls out the differences between blocks and views, but that's a constant and this has taken me too long already. Worth thinking about for the future.

### CODE REVIEW

**Comment by 86667** [comment]
Very nice separation and thanks for the docs. Conceptually looks good my only comments right now are minor:

- Is there a reason you've chosen `ExternalMessage::ProcessProposal` to be an external message? It feels Internal to me and would make the code more intuitive to understand at first glance.
- There is usage of "tail" and "head" to mean the same thing, which you mention in comments. Shall we pick one?

**Comment by rrw-zilliqa** [comment]
> Very nice separation and thanks for the docs. Conceptually looks good my only comments right now are minor:
> 
> * Is there a reason you've chosen `ExternalMessage::ProcessProposal` to be an external message? It feels Internal to me and would make the code more intuitive to understand at first glance.

I didn't know what the right thing to do was there and so I did what exists and worked. Happy to do whatever turns out to be best .. (previously, we sent Proposal ExternalMessage's to ourselves, and now we send ProcessProposals on the same basis).

> * There is usage of "tail" and "head" to mean the same thing, which you mention in comments. Shall we pick one?

Go for it :-)

File: zilliqa/src/block_store.rs [hunk 1]
```diff
@@ -1,25 +1,333 @@
...
+        u64::try_from(key >> self.shift).unwrap()
+    }
+
+    pub fn min_key_for_view(&self, view: u64) -> u128 {
+        u128::from(view) << self.shift
+    }
+
+    pub fn min_tail_key(&self, highest_known_view: u64) -> u128 {
+        let highest_key = u128::from(highest_known_view + 1) << self.shift;
+        highest_key - u128::try_from(constants::BLOCK_CACHE_TAIL_BUFFER_ENTRIES).unwrap()
```
Comment (86667) [inline zilliqa/src/block_store.rs:105]: What is the purpose of taking `BLOCK_CACHE_TAIL_BUFFER_ENTRIES` off this value?

Comment (rrw-zilliqa) [reply zilliqa/src/block_store.rs:105]: This function returns the minimum key ( view_number << log_2(ways) ) which will be retained in the tail cache.
highest_key is the highest key we think might ever exist in the system.
One subtracts the number of entries in the tail cache from the highest key to find the key above which we are prepared to store entries in the tail cache.
Subtract too much and the tail cache will become larger than you want. Subtract too little and it will become smaller - subtract nothing and the only entries you will be prepared to store are those that don't exist yet.

Comment (rrw-zilliqa) [reply zilliqa/src/block_store.rs:105]: This function returns the minimum key ( view_number << log_2(ways) ) which will be retained in the tail cache.
highest_key is the highest key we think might ever exist in the system.
One subtracts the number of entries in the tail cache from the highest key to find the key above which we are prepared to store entries in the tail cache.
Subtract too much and the tail cache will become larger than you want. Subtract too little and it will become smaller - subtract nothing and the only entries you will be prepared to store are those that don't exist yet.

(I'll add a comment in the code :-)  - thanks for pointing out!)

Comment (rrw-zilliqa) [reply zilliqa/src/block_store.rs:105]: This function returns the minimum key ( view_number << log_2(ways) ) which will be retained in the tail cache.
highest_key is the highest key we think might ever exist in the system.
One subtracts the number of entries in the tail cache from the highest key to find the key above which we are prepared to store entries in the tail cache.
Subtract too much and the tail cache will become larger than you want. Subtract too little and it will become smaller - subtract nothing and the only entries you will be prepared to store are those that don't exist yet.

(I'll add a comment in the code :-)  - thanks for pointing out!)

Comment (86667) [reply zilliqa/src/block_store.rs:113]: Thanks! Understood. Deriving this from `self.highest_known_view` (which is set during in `buffer_proposal` just before we insert), means that the `tail` will only contain proposals that are >= the largest view we've seen. So if we are block 100 and syncing, and we receive new block 100_000, we put only proposals with view >= 100_000 in `tail`. 

So this acts as a mechanism to buffer new activity happening at the head whilst connected to the network but syncing. The size of this cache should then depend on sync times - if in sync time there are 1000 new proposals at the head then this cache doesnt do much.

Please correct me if that isnt accurate! I'll propose changing the name to `look_ahead_cache` or something in a follow up PR

Comment (rrw-zilliqa) [reply zilliqa/src/block_store.rs:113]: Yes - it's intended to cure a problem where the head moves on as you are fetching the blocks up to it. Two things then happen - first, as you say, you are then throwing away broadcast proposals so as to fetch them later, which is wasteful (but the main cache ought to catch them if you really wanted them[1]), and second, by the time you've received a block as a result of one of your requests, the head has moved on and so there's another block to fetch .. 

As such, the only capacity you really need here is one RTT (maybe two)  - at one block per second 16 entries gives you 15 or so seconds, which I agree may not be enough, but I thought it was!

If you're interested in buffering blocks generally, the main block cache should do that for you, advancing its lower bound as you add blocks to your chain.

I'd go for `head_cache` or similar (`look_ahead_cache` kinda implies we're looking ahead of the chain, which we're not - though that is another way to solve the problem: remove this cache and just have the main cache, and simply request views ahead of where you are in the belief that they probably exist).

[1] I took the view that you shouldn't attempt to fix this problem too hard in the tail cache, because you only have a limited amount of memory space, and if you're after enough space to buffer the gap between the currently broadcast block and the last synced block, you should allocate it to the main buffer cache.

**Review by JamesHinshelwood — APPROVED** [review]
Thanks and apologies again for being so slow to review!

A bunch of comments, but I don't think anything is a major blocker so feel free to:
* Reply to/address comments here
* Address comments in a follow up
* Ignore me because I'm wrong

File: docs/fetching_blocks.md
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply docs/fetching_blocks.md:1]: I can certainly move the docs :-) - where would you like them? The top of block_store.rs ?

File: zilliqa/src/consensus.rs [hunk 2]
```diff
@@ -2026,6 +2025,47 @@ impl Consensus {
         Ok(())
     }
 
+    /// Trigger a checkpoint, for debugging.
+    /// Returns (file_name, block_hash). At some time after you call this function, hopefully a checkpoint will end up in the file
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/consensus.rs:2029]: I agree with you entirely! Happy to take action if you want .. (I keep the debug API out of tree on a branch, for this reason)

File: zilliqa/src/message.rs [hunk 3]
```diff
@@ -196,6 +212,18 @@ pub struct BlockRequest {
...
+    /// When we send a block response, we may also send data on what blocks we are prepared
+    /// to serve.
+    pub availability: Option<Vec<BlockStrategy>>,
+}
+
+/// Used to convey proposal processing internally, to avoid blocking threads for too long.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ProcessProposal {
+    // An encoded PeerId
+    pub from: Vec<u8>,
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/message.rs:225]: Because I wanted to force the representation of fear of incompatibility if PeerId (which is in an external library) changed its external representation; very happy to change it?

File: zilliqa/src/message.rs [hunk 4]
```diff
@@ -196,6 +212,18 @@ pub struct BlockRequest {
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct BlockResponse {
     pub proposals: Vec<Proposal>,
+    pub from_view: u64,
+    /// When we send a block response, we may also send data on what blocks we are prepared
+    /// to serve.
+    pub availability: Option<Vec<BlockStrategy>>,
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/message.rs:218]: I'll add a comment - the former means "I don't know what the availability is" and the latter means "I know that I know nothing"

File: docs/fetching_blocks.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+Also, the maximum depth of a fork is the maximum lookahead of our
+cache. Suppose we have a fork 100 deep starting at view 202. We'll
+fetch it, go to the end, and stop.
+
+How would we ever know that there was another block whose parent was
+the block at view 200? Well, unless we ask for subsequent blocks, we
+won't. And we won't ask for subsequent blocks because we're too busy
+re-requesting the ones we can't process.
+
+This is theoretically fixable and we should one day - but I've not
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply docs/fetching_blocks.md:155]: Only if the node that's telling you about them is honest. If it's lying, then it can construct forks of arbitrary depth and if they are long enough, you may not be able to see past them to pick up the true chain again :-(

File: docs/fetching_blocks.md
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply docs/fetching_blocks.md:1]: I've pulled out all the docs for individual functions and put them alongside the functions, added a reference to the overview document in the tick function where it might be useful. Thanks for the hint - I hope that was what you were expecting?

File: docs/fetching_blocks.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+ - Other nodes will, in general, not have all the blocks themselves -
+   they may be syncing, or have pruned some blocks for resource
+   management reasons.
+ - We want to avoid overloading a single node with too much work.
+ - The syncing node wants to catch up quickly.
+
+As a result, the code for fetching blocks to complete our view of the
+chain is somewhat complex. It consists of:
+
+For essentially historical reasons, blocks are fetched by view, not block number.
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply docs/fetching_blocks.md]: I think we could talk about this for a while, but I've rephrased in a way I hope means we don't need to go down that rabbit hole :-)

File: zilliqa/src/consensus.rs [hunk 7]
```diff
@@ -2191,6 +2231,21 @@ impl Consensus {
         Ok(())
     }
 
+    // Receives availability and passes it on to the block store.
+    pub fn receive_availability(
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/consensus.rs]: nit picked

File: zilliqa/src/consensus.rs [hunk 8]
```diff
@@ -2219,14 +2274,21 @@ impl Consensus {
...
-                }),
-            )?;
-        }
+        // We may have child blocks; process them too.
+        self.block_store
+            .process_block(from, block)?
+            .into_iter()
+            .for_each(|(from_id, child_proposal)| {
+                // If we fail here, just fail - eventually someone will notice we did and
+                // retry the whole action.
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/consensus.rs:2310]: OK - that's a fair comment; originally, because if you fail here for one block you will throw all the others away and if that failure was because a buffer was full or similar, you might well end up in a repeating loop. Just let's remember not to make send_external_message() fail for another reason!

File: zilliqa/src/message.rs [hunk 9]
```diff
@@ -217,6 +245,7 @@ pub enum ExternalMessage {
     NewView(Box<NewView>),
     BlockRequest(BlockRequest),
     BlockResponse(BlockResponse),
+    ProcessProposal(ProcessProposal),
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/message.rs:248]: We trust ProcessProposal because we sent it. We don't trust Proposal. Feel free to suggest a better/different name.

File: zilliqa/src/consensus.rs [hunk 10]
```diff
@@ -2808,4 +2867,56 @@ impl Consensus {
     ) -> Result<()> {
         self.block_store.report_outgoing_message_failure(failure)
     }
+
+    pub fn tick(&mut self) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/consensus.rs:2889]: Because I wasn't sure if you wanted to regard a process normally executed periodically as a "timeout" .

Necessity, really - we generate traffic when we trigger block requests; I couldn't think of a way for a message to trigger these without risking generating either too much traffic (and blocking the node - and others - from other work, which was my initial problem), or too little traffic (and never requesting the blocks it needed to complete its view of the chain). So, trigger on a tick ..

File: zilliqa/src/block_store.rs [hunk 11]
```diff
@@ -112,57 +555,91 @@ impl BlockStore {
...
+    /// We need to do this by view range, which means that we need to account for views where there was no block.
+    /// So, the underlying db function finds the view lower and upper bounds of our contiguous block ranges and we
+    /// advertise those.
+    /// This function overlays that with a timeout so we don't ask (and thus load the db) too often.
+    pub fn availability(&mut self) -> Result<Option<Vec<BlockStrategy>>> {
+        let mut to_return = self.strategies.clone();
+        let now = SystemTime::now();
+        if self.available_blocks_updated.map_or(true, |x| {
+            now.duration_since(x).unwrap_or(Duration::ZERO)
+                > Duration::from_secs(constants::RECOMPUTE_BLOCK_AVAILABILITY_AFTER_S)
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/block_store.rs:603]: yes, absolutely - but this is a fairly expensive db query, so timer triggered to avoid doing it too often and causing too great a load; this way, it gets done eventually, but we bound the amount of work we are prepared to do to keep it up to date. Smarter algorithms certainly possible, just ... harder :-)

[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/block_store.rs:603]: (in particular, guards against block proposal floods stalling the node by causing it to continually recompute availability; could as easily as you say keep this in memory and recompute this on block insert rather than availability query - however, that would've been a bit more intrusive, so I didn't)

File: zilliqa/src/constants.rs
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/constants.rs:1]: Good point well made :-) - fixed.

File: zilliqa/src/consensus.rs [hunk 12]
```diff
@@ -2808,4 +2867,56 @@ impl Consensus {
...
+        }
+        Ok(())
+    }
+
+    pub fn buffer_proposal(&mut self, from: PeerId, proposal: Proposal) -> Result<()> {
+        self.block_store.buffer_proposal(from, proposal)?;
+        Ok(())
+    }
+
+    pub fn buffer_lack_of_proposals(
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/consensus.rs:2935]: Added documentation - hope that makes things clearer ... ?

File: zilliqa/src/constants.rs [hunk 13]
```diff
@@ -26,3 +26,36 @@ pub const ZIL_CONTRACT_CREATE_GAS: usize = 50;
...
+
+// log2 of the number of ways in the block cache. Max 8.
+pub const BLOCK_CACHE_LOG2_WAYS: usize = 4;
+/// The block cache will deliberately keep this number of entries near the highest known view
+/// so that it can catch up quickly when it reaches the current head of the chain.
+/// Not required for correctness.
+pub const BLOCK_CACHE_TAIL_BUFFER_ENTRIES: usize = 16;
+
+/// Max pending requests per peer
+pub const MAX_PENDING_REQUESTS_PER_PEER: usize = 4;
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/constants.rs]: Name changed to be more descriptive (I hope!)

File: zilliqa/src/constants.rs [hunk 14]
```diff
@@ -26,3 +26,36 @@ pub const ZIL_CONTRACT_CREATE_GAS: usize = 50;
...
+pub const RECOMPUTE_BLOCK_AVAILABILITY_AFTER_S: u64 = 2;
+
+// Maximum rate at which to send availability requests
+pub const REQUEST_PEER_VIEW_AVAILABILITY_NOT_BEFORE_MS: u64 = 1000;
+
+// We assume that every node has the last ALWAYS_RETAIN_LAST_N_BLOCKS blocks, otherwise
+// it's hard ever to catch up. Set this too large and syncing will be hard because we will
+// keep asking recently started nodes for blocks they don't have. Set it too small and
+// it will be hard because we'll need to keep waiting for availability.
+pub const RETAINS_LAST_N_BLOCKS: u64 = 10;
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/constants.rs:47]: Added comment - 100 might lead to us requesting blocks that the node doesn't have, which will lead to slow syncs and timeouts.

File: zilliqa/src/db.rs [hunk 15]
```diff
@@ -778,6 +799,48 @@ impl Db {
...
     }
+
+    /// Retrieve a list of the views in our db.
+    /// This is a bit horrific. What we actually do here is to find the view lower and upper bounds for the contiguous block ranges in the database.
+    /// See block_store.rs::availability() for details.
+    pub fn get_view_ranges(&self) -> Result<Vec<Range<u64>>> {
+        // The island field is technically redundant, but it helps with debugging.
+        Ok(self.block_store.lock().unwrap()
+            .prepare_cached("SELECT MIN(vlb), MAX(vub), MIN(height),MAX(height),height-rank AS island FROM ( SELECT height,vlb,vub,ROW_NUMBER() OVER (ORDER BY height) AS rank FROM
+ (SELECT height,MIN(view) as vlb, MAX(view) as vub from blocks GROUP BY height ) )  GROUP BY island ORDER BY MIN(height) ASC")?
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/db.rs:830]: Added an explanatory comment, if that helps?

File: zilliqa/src/db.rs [hunk 16]
```diff
@@ -778,6 +799,48 @@ impl Db {
...
+           .query_map([], Self::make_view_range)?.collect::<Result<Vec<_>,_>>()?)
+    }
+
+    /// Forget about a range of blocks; this saves space, but also allows us to test our block fetch algorithm.
+    /// If canonical is true, we'll forget the canonical block mappings for this range too - uses less space, but not so good for security.
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/db.rs:835]: I did. Fixed.

File: zilliqa/src/db.rs [hunk 17]
```diff
@@ -778,6 +799,48 @@ impl Db {
...
+    pub fn forget_block_range(&self, blocks: Range<u64>) -> Result<()> {
+        self.with_sqlite_tx(move |tx| {
+            // Remove everything!
+            tx.execute("DELETE FROM tip_info WHERE latest_finalized_view IN (SELECT view FROM blocks WHERE height >= :low AND height < :high)",
+                       named_params! {
+                           ":low" : blocks.start,
+                           ":high" : blocks.end } )?;
+            // @TODO can't yet remove transactions - we don't know the hashes.
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/src/db.rs:843]: thanks! Comment removed.

File: zilliqa/tests/it/main.rs [hunk 18]
```diff
@@ -749,6 +753,19 @@ impl Network {
         // Advance time.
         zilliqa::time::advance(Duration::from_millis(1));
 
+        // Every 10ms send a consensus tick, since most of our tests are too short to otherwise
+        // be able to sync.
```
[the remark this answers is not in the document]
Comment (rrw-zilliqa) [reply zilliqa/tests/it/main.rs:757]: because it only fires (I think!) when there are no more messages to send, which never happens for some tests...

File: docs/fetching_blocks.md
[the remark this answers is not in the document]
Comment (JamesHinshelwood) [reply docs/fetching_blocks.md:1]: Awesome, thank you, sorry for nitpicking but much appreciated.

File: zilliqa/src/consensus.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (JamesHinshelwood) [reply zilliqa/src/consensus.rs:2061]: I'll raise an issue, thanks :)

File: zilliqa/src/consensus.rs [hunk 19]
```diff
@@ -2838,4 +2899,58 @@ impl Consensus {
...
+                self.block_store.delete_empty_view_range_cache();
+            } else {
+                likely_blocks.into_iter().for_each(|(from, block)| {
+                    trace!(
+                        "buffer may contain the next block - {0:?} v={1} n={2}",
+                        block.hash(),
+                        block.view(),
+                        block.number()
+                    );
+                    // Ignore errors here - just carry on and wait for re-request to clean up.
```
Comment (JamesHinshelwood) [inline zilliqa/src/consensus.rs:2927]: There's another ignored `Err` here that I would argue we should propagate up for the same reason as previously.

Happy to do that in a separate PR though - As I said, the node is in the process of  to explode if this would have return an `Err` anyhow

**Review by JamesHinshelwood — APPROVED** [review]
Thanks! Lets see how this does on the devnet :)

**Review by JamesHinshelwood — APPROVED** [review]

File: zilliqa/src/consensus.rs (hunk 19 above)
Comment (rrw-zilliqa) [reply zilliqa/src/consensus.rs:2918]: @JamesHinshelwood  - sounds good to me; since I'm in the area, would you prefer:

 1. Try to send all the messages and return the first error (if there is one), or
 2. Send messages until  you hit an error, then return it.
 
 ?

**Comment by rrw-zilliqa** [comment]
This PR has been superceded by https://github.com/Zilliqa/zq2/pull/1680
